src/libraries/System.Text.Json/src/System/Text/Json/Serialization/ReflectionEmitMemberAccessor.cs

```diff
@@ -246,40 +260,92 @@ private static DynamicMethod CreatePropertyGetter(PropertyInfo propertyInfo, Typ
+        {
+            string memberName = propertyInfo.Name;
+            Type declaredPropertyType = propertyInfo.PropertyType;
+            Type runtimePropertyType = typeof(TProperty);
+
+            DynamicMethod propertySetter = CreatePropertySetter(propertyInfo, memberName, declaredPropertyType, runtimePropertyType);
 
-        private static DynamicMethod CreatePropertySetter(PropertyInfo propertyInfo, Type propertyType)
+            if (declaredPropertyType == runtimePropertyType)
+            {
+                // If declared and runtime type are identical, the assignment cannot fail, thus
+                // there is no need for exception handling or closures.
+
+                return CreateDelegate<Action<object?, TProperty>>(propertySetter);
+            }
+            else
+            {
+                // Return a delegate with some closures needed for creating the potential exception.
+
+                Action<object?, TProperty, string, Type> propertySetterDelegate =
+                    CreateDelegate<Action<object?, TProperty, string, Type>>(propertySetter);
+
+                return (object? obj, TProperty value) => propertySetterDelegate(obj, value, memberName, declaredPropertyType);
+            }
+        }
+
+        private static DynamicMethod CreatePropertySetter(PropertyInfo propertyInfo, string memberName, Type declaredPropertyType, Type runtimePropertyType)
         {
             MethodInfo? realMethod = propertyInfo.SetMethod;
             Debug.Assert(realMethod != null);
 
             Type? declaringType = propertyInfo.DeclaringType;
             Debug.Assert(declaringType != null);
 
-            DynamicMethod dynamicMethod = CreateSetterMethod(propertyInfo.Name, propertyType);
-            ILGenerator generator = dynamicMethod.GetILGenerator();
+            // declaredPropertyType: Type of the property
+            // runtimePropertyType:  <T> of JsonConverter / JsonPropertyInfo
 
-            generator.Emit(OpCodes.Ldarg_0);
+            bool valueNeedsCastOrUnbox = declaredPropertyType != runtimePropertyType;
 
-            if (declaringType.IsValueType)
+            DynamicMethod dynamicMethod = CreateSetterMethod(memberName, runtimePropertyType, valueNeedsCastOrUnbox);
+            ILGenerator generator = dynamicMethod.GetILGenerator();
+
+            if (!valueNeedsCastOrUnbox)
             {
-                generator.Emit(OpCodes.Unbox, declaringType);
+                generator.Emit(OpCodes.Ldarg_0);
+                generator.Emit(declaringType.IsValueType ? OpCodes.Unbox : OpCodes.Castclass, declaringType);
                 generator.Emit(OpCodes.Ldarg_1);
-                generator.Emit(OpCodes.Call, realMethod);
             }
             else
             {
-                generator.Emit(OpCodes.Castclass, declaringType);
+                LocalBuilder value = generator.DeclareLocal(declaredPropertyType);
+
+                // try
```

Comment:
I don't think adding a try\catch is the right thing to do. Letting the raw exception (InvalidCast; perhaps NullRefeference) propagate due to an invalid converter is acceptable IMO.

A try\catch is also not completely free from a perf standpoint (even when no exception is thrown) and doing that on every setter\getter would likely impact perf.

Reply:
@steveharter I think you are reviewing an outdated version. All that stuff is in a separate PR #41323 now